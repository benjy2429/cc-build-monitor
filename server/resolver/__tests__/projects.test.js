import resolveProjects from '../projects';
import projectMock from '../../../mocks/projects/normal';
import projectMockMalformed from '../../../mocks/projects/malformed';
import projectMockEmpty from '../../../mocks/projects/empty';

describe('Resolve projects', () => {
  it('parses xml data', async () => {
    const mockFetch = () => Promise.resolve(projectMock);
    const result = await resolveProjects(mockFetch);
    expect(result).toMatchSnapshot();
  });

  it('returns an empty array when no data is fetched', async () => {
    const mockFetch = () => Promise.resolve(projectMockEmpty);
    const result = await resolveProjects(mockFetch);
    expect(result).toMatchSnapshot();
  });

  it('returns an error when the xml is malformed', async () => {
    const mockFetch = () => Promise.resolve(projectMockMalformed);
    const result = await resolveProjects(mockFetch);
    expect(result).toMatchSnapshot();
  });

  it('returns an error when the fetch fails', async () => {
    const mockFetch = () => Promise.reject({ message: 'an error' });
    const result = await resolveProjects(mockFetch);
    expect(result).toMatchSnapshot();
  });

  it('excludes projects in the blacklist', async () => {
    const mockConfig = { blacklist: ['project1'], whitelist: [] };
    const mockFetch = () => Promise.resolve(projectMock);
    const result = await resolveProjects(mockFetch, mockConfig);
    expect(result).toMatchSnapshot();
  });

  it('only includes projects in the whitelist', async () => {
    const mockConfig = { blacklist: [], whitelist: ['project1'] };
    const mockFetch = () => Promise.resolve(projectMock);
    const result = await resolveProjects(mockFetch, mockConfig);
    expect(result).toMatchSnapshot();
  });

  it('prioritises the whitelist over the blacklist', async () => {
    const mockConfig = { blacklist: ['project1'], whitelist: ['project2'] };
    const mockFetch = () => Promise.resolve(projectMock);
    const result = await resolveProjects(mockFetch, mockConfig);
    expect(result).toMatchSnapshot();
  });
});
