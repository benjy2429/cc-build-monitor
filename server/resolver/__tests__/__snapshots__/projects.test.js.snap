// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resolve projects parses xml data 1`] = `
Object {
  "projects": Array [
    Object {
      "activity": "Sleeping",
      "lastBuildLabel": "1",
      "lastBuildStatus": "Success",
      "lastBuildTime": "2000-01-01T12:00:00.000Z",
      "name": "project1",
    },
    Object {
      "activity": "Sleeping",
      "lastBuildLabel": "13",
      "lastBuildStatus": "Failure",
      "lastBuildTime": "2000-01-01T12:00:00.000Z",
      "name": "project2",
    },
    Object {
      "activity": "Building",
      "lastBuildLabel": "42",
      "lastBuildStatus": "Unknown",
      "lastBuildTime": "2000-01-01T12:00:00.000Z",
      "name": "project3",
    },
  ],
}
`;

exports[`Resolve projects returns an empty array when no data is fetched 1`] = `
Object {
  "projects": Array [],
}
`;

exports[`Resolve projects returns an error when the fetch fails 1`] = `
Object {
  "error": "an error",
}
`;

exports[`Resolve projects returns an error when the xml is malformed 1`] = `
Object {
  "error": "Error parsing response",
}
`;
